kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,188 @@
+pub fn extract_unsize_casting<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    src_ty: Ty<'tcx>,
+    dst_ty: Ty<'tcx>,
+) -> (Ty<'tcx>, Ty<'tcx>) {
+    trace!(?dst_ty, ?src_ty, "find_trait_conversion");
+    let info = CoerceUnsizedIterator::new(tcx, src_ty, dst_ty).last().unwrap();
+    let src_base_ty = extract_pointee(info.src_ty)
+        .expect(&format!("Expected source to be a pointer. Found {:?} instead", info.src_ty));
+    let dst_base_ty = extract_pointee(info.dst_ty)
+        .expect(&format!("Expected destination to be a pointer. Found {:?} instead", info.dst_ty));
+    tcx.struct_lockstep_tails_erasing_lifetimes(src_base_ty, dst_base_ty, ParamEnv::reveal_all())
+}
+
+/// Iterates over the coercion path of a structure that implements `CoerceUnsized<T>` trait.
+/// The `CoerceUnsized<T>` trait indicates that this is a pointer or a wrapper for one, where
+/// unsizing can be performed on the pointee. More details:
+/// <https://doc.rust-lang.org/std/ops/trait.CoerceUnsized.html>
+///
+/// Given an unsized coercion between `impl CoerceUnsized<T>` to `impl CoerceUnsized<U>` where
+/// `T` is sized and `U` is unsized, this iterator will walk over the fields that lead to a
+/// pointer to `T`, which shall be converted from a thin pointer to a fat pointer.
+///
+/// This iterator will also include the name of the field that differs in the current iteration.
+///
+/// The first element of the iteration will always be the starting types.
+/// The last element of the iteration will always be pointers to `T` and `U`.
+/// After unsized element has been found, the iterator will return `None`.
+pub struct CoerceUnsizedIterator<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    src_ty: Option<Ty<'tcx>>,
+    dst_ty: Option<Ty<'tcx>>,
+}
+
+/// Represent the information about a coercion.
+#[derive(Debug, Clone, PartialOrd, PartialEq)]
+pub struct CoerceUnsizedInfo<'tcx> {
+    /// The name of the field from the current types that differs between each other.
+    pub field: Option<Symbol>,
+    /// The type being coerced.
+    pub src_ty: Ty<'tcx>,
+    /// The type that is the result of the coercion.
+    pub dst_ty: Ty<'tcx>,
+}
+
+impl<'tcx> CoerceUnsizedIterator<'tcx> {
+    pub fn new(
+        tcx: TyCtxt<'tcx>,
+        src_ty: Ty<'tcx>,
+        dst_ty: Ty<'tcx>,
+    ) -> CoerceUnsizedIterator<'tcx> {
+        CoerceUnsizedIterator { tcx, src_ty: Some(src_ty), dst_ty: Some(dst_ty) }
+    }
+}
+
+/// Iterate over the coercion path. At each iteration, it returns the name of the field that must
+/// be coerced, as well as the current source and the destination.
+/// E.g.: The first iteration of casting `Rc<String>` -> `Rc<&dyn Debug>` will return
+/// ```rust,ignore
+/// CoerceUnsizedInfo { field: "ptr", src_ty /* Rc<String> */, dst_ty /* Rc<&dyn Debug> */}
+/// ```
+/// while the last iteration will return:
+/// ```rust,ignore
+/// CoerceUnsizedInfo {
+///   field: None,
+///   src_ty: Ty, /* *const RcBox<String> */,
```

Comment:
So it stops at the box, and doesn't go inside?  I guess that makes some sense cause the box is really a pointer?

Reply:
This is not a `Box` though, it's an `RcBox` which is the internal structure of `Rc`.

Reply:
I changed to example to use `NonNull`, so hopefully it's a bit easier to understand.